src/librustc/mir/interpret/mod.rs

```diff
@@ -195,7 +194,21 @@ where
             assert!(cache(encoder).insert(alloc_id, pos).is_none());
             trace!("encoding {:?} with {:#?}", alloc_id, alloc);
             AllocKind::Alloc.encode(encoder)?;
+
+            // Write placeholder for size
+            let size_pos = encoder.position();
+            0usize.encode(encoder)?;
```

Comment:
This doesn't work because of variable-length integer encoding.

Reply:
rustc has some similar code elsewhere and works around this by using a 4 byte array that the size is encoded into. Somewhat space-wasteful though.

Reply:
It might be simpler to just remember the in the `global_cache` during decoding?

Reply:
that doesn't work, because we need this value also when another thread hasn't finished decoding the allocation yet.